docs(flags): Property overrides documentation (#14309)
PostHog/posthog.com: contents/docs/feature-flags/property-overrides.mdx

sidebar: Docs
showTitle: true
availability:
    free: full
    selfServe: full
    enterprise: full
---

Property overrides let you provide person or group properties directly to PostHog for feature flag evaluation, instead of relying on properties stored on the server. This is useful when properties are set from a different client (like a backend service setting a property that your frontend needs immediately), or when multiple clients setting the same property could cause inconsistent flag evaluations.

## Why use property overrides?

Property overrides are useful when:

1. **Properties are set elsewhere**: Properties might be set from different parts of your application (like a backend service or another client), and your frontend doesn't know if they've been processed. Additionally, when multiple clients set the same property, the value in PostHog can also change unpredictably as updates arrive from different sources.

2. **Values are transient or context-specific**: You may need to evaluate flags based on values that shouldn't be stored on the person record, such as session-specific context, derived values, or data you'd prefer not to persist.

> 3. **Testing and debugging**: You may want to test flag behavior with specific property values without modifying the actual user's properties.

Reviewer: ```suggestion
3. **Testing and debugging**: You can test flag behavior using specific property values. This way, you won't need to change the actual user's properties.
```

```

```react-native
posthog.group('company', 'company-123', {
    name: 'Acme Inc',
    plan: 'enterprise'
})

// Group properties are available for flag evaluation
const showFeature = posthog.isFeatureEnabled('enterprise-companies-feature')
```

```swift file=iOS
PostHogSDK.shared.group(type: "company", key: "company-123", groupProperties: [
    "name": "Acme Inc",
    "plan": "enterprise"
])

// Group properties are available for flag evaluation
let showFeature = PostHogSDK.shared.isFeatureEnabled("enterprise-companies-feature")
```

```kotlin file=Android
PostHog.group(
    "company",
    "company-123",
    groupProperties = mapOf(
        "name" to "Acme Inc",
        "plan" to "enterprise"
    )
)

// Group properties are available for flag evaluation
val showFeature = PostHog.isFeatureEnabled("enterprise-companies-feature")
```

</MultiLanguage>

## Default person properties

Mobile SDKs (React Native, iOS, and Android) can automatically include common device and app properties in every feature flag evaluation request. This ensures flags targeting properties like app version or OS work correctly without waiting for server-side processing.

> This behavior is controlled by the `setDefaultPersonProperties` configuration option, which defaults to `true`.

Reviewer: ```suggestion
The `setDefaultPersonProperties` configuration option controls this behavior and defaults to `true`.
```

```

```kotlin file=Android
// Set group properties for flag evaluation
PostHog.setGroupPropertiesForFlags(
    "company",
    mapOf(
        "plan" to "enterprise",
        "employee_count" to 500
    )
)

// Reset group properties for a specific type
PostHog.resetGroupPropertiesForFlags("company")

// Reset all group properties
PostHog.resetGroupPropertiesForFlags()
```

</MultiLanguage>

### Use case: Targeting by app version

With `setDefaultPersonProperties` enabled, you can create feature flags that target specific app versions without any additional code:

1. Create a feature flag in PostHog
2. Add a filter condition: `$app_version` equals `2.0.0`
3. The flag automatically matches users on version 2.0.0

This works immediately because the SDK sends the app version with every flag request.

## GeoIP property overrides

PostHog's servers automatically derive GeoIP properties from the user's IP address for feature flag evaluation. This enables geolocation-based flags without any manual setup on client-side SDKs.

### Properties included

| Property | Description |
|----------|-------------|
| `$geoip_city_name` | City name |
| `$geoip_country_name` | Country name |

> | `$geoip_country_code` | Two-letter country code |

Reviewer 2: If we had (I couldn't find with a quick search though) a table with all country codes it would be nice to link them here.

        PersonProperties: posthog.NewProperties().
            Set("plan", "enterprise").
            Set("company_size", "large"),
        Groups: posthog.NewGroups().
            Set("company", "company-123"),
        GroupProperties: map[string]map[string]interface{}{
            "company": {
                "plan":           "enterprise",
                "employee_count": 500,
            },
        },
    },
)
```

```java
Object flagValue = posthog.getFeatureFlag(
    "user-distinct-id",
    "feature-flag-key",
    PostHogFeatureFlagOptions.builder()
        .groups(Map.of("company", "company-123"))
        .personProperties(Map.of(
            "plan", "enterprise",
            "company_size", "large"
        ))
        .groupProperties(Map.of(
            "company", Map.of(
                "plan", "enterprise",
                "employee_count", 500
            )
        ))
        .build()
);
```

</MultiLanguage>

### GeoIP on server-side

Server-side SDKs disable GeoIP by default. This is because PostHog derives GeoIP properties from the request's IP address, which for server-side requests is your server's IP rather than your user's.

> For geolocation-based flags, evaluate them on the client side where the user's actual IP is available.

Reviewer 2: When the backend has the user's IP address, isn't it possible to pass it on server-side SDKs?
From this paragraph, my understanding is that it isn't possible, yet we say the is disabled "by default", I would rephrase that to be clear that it isn't possible or quickly explain how the user could do that (if we have a link to it, maybe simply adding it would suffice).

Author: I purposely omitted this, at least for now, because we don't have a great way of exposing this. In some SDKs (Python, Node), GeoIP can be enabled, but the flags service looks at the IP sent in the request headers (e.g., `x-real-ip`, `x-forwarded-for`), not anything provided in the request body. So by re-enabling it, you'll always get the IP of the server.

It may be technically possible to customize the headers sent by the HTTP client, but I'm hesitant to call that out here due to the fact that it's not really supported and would vary heavily by language.

I think that, if anything, this is a good potential enhancement for our server-side SDKs in the future.
